### ISSUE TITLE
Compiler errors on macOS

### ISSUE THREAD

**Issue by doberkofler** [NectarJS/nectarjs#66]
I love the idea of a native JavaScript compiler but cannot seem to get my hello world running. What am I doing wrong?

```bash
gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.3 (clang-1103.0.32.59)
Target: x86_64-apple-darwin19.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

```js
console.log("hello world");
```

```bash
npx nectar helloworld.js
[*] Generating source file
[*] Compiling
In file included from /Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp:42:
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:68:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>> __OBJECT;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:115:22: error: expected ';' after top level declarator
val REGISTER[1000000]{(val){.i=0}};
                     ^
                     ;
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:610:95: error: a space is required between consecutive right angle brackets (use '> >')
inline __NJS_VAR __NJS_Object_Set(char* _index, __NJS_VAR _value, vector<pair<char*, __NJS_VAR>>* _obj)
                                                                                              ^~
                                                                                              > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:631:99: error: a space is required between consecutive right angle brackets (use '> >')
inline __NJS_VAR __NJS_Object_Set(__NJS_VAR _index, __NJS_VAR _value, vector<pair<char*, __NJS_VAR>>* _obj)
                                                                                                  ^~
                                                                                                  > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:652:31: error: a space is required between consecutive right angle brackets (use '> >')
  vector<pair<char*, __NJS_VAR>>* _obj;
                              ^~
                              > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:693:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>>* _obj;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:760:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>>* _obj;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:801:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>>* _obj;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:827:78: error: expected expression
  function<__NJS_VAR ()>* __OBJ_TO___NJS_STRING = new function<__NJS_VAR ()>([&](){ return __NJS_Create_String((char*)this->__NJS_VALUE.c_str()); });
                                                                             ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:837:95: error: expected expression
  function<__NJS_VAR (__NJS_VAR)>* __OBJ_TO___NJS_SPLIT = new function<__NJS_VAR (__NJS_VAR)>([&](__NJS_VAR _needle)
                                                                                              ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:877:78: error: expected expression
  function<__NJS_VAR ()>* __OBJ_TO___NJS_STRING = new function<__NJS_VAR ()>([&](){ return __NJS_Create_String((char*)"Array"); });
                                                                             ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp:56:95: error: expected expression
        __NJS_DECL_FUNCTION<__NJS_VAR ()>* __NJS_FN___fc534r = new __NJS_DECL_FUNCTION<__NJS_VAR ()>([&]() -> __NJS_VAR{
                                                                                                     ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp:81:94: error: expected expression
__NJS_DECL_FUNCTION<__NJS_VAR ()>* __NJS_FN___2l38vg = new __NJS_DECL_FUNCTION<__NJS_VAR ()>([&]() -> __NJS_VAR{
                                                                                             ^
13 errors generated.
[!] Compilation error
Error: Command failed: g++ /Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp  -O1 -fpermissive -w -s  -o /Users/doberkofler/MyDev/doberkofler/nectarjs-test/helloworld
In file included from /Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp:42:
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:68:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>> __OBJECT;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:115:22: error: expected ';' after top level declarator
val REGISTER[1000000]{(val){.i=0}};
                     ^
                     ;
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:610:95: error: a space is required between consecutive right angle brackets (use '> >')
inline __NJS_VAR __NJS_Object_Set(char* _index, __NJS_VAR _value, vector<pair<char*, __NJS_VAR>>* _obj)
                                                                                              ^~
                                                                                              > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:631:99: error: a space is required between consecutive right angle brackets (use '> >')
inline __NJS_VAR __NJS_Object_Set(__NJS_VAR _index, __NJS_VAR _value, vector<pair<char*, __NJS_VAR>>* _obj)
                                                                                                  ^~
                                                                                                  > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:652:31: error: a space is required between consecutive right angle brackets (use '> >')
  vector<pair<char*, __NJS_VAR>>* _obj;
                              ^~
                              > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:693:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>>* _obj;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:760:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>>* _obj;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:801:33: error: a space is required between consecutive right angle brackets (use '> >')
    vector<pair<char*, __NJS_VAR>>* _obj;
                                ^~
                                > >
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:827:78: error: expected expression
  function<__NJS_VAR ()>* __OBJ_TO___NJS_STRING = new function<__NJS_VAR ()>([&](){ return __NJS_Create_String((char*)this->__NJS_VALUE.c_str()); });
                                                                             ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:837:95: error: expected expression
  function<__NJS_VAR (__NJS_VAR)>* __OBJ_TO___NJS_SPLIT = new function<__NJS_VAR (__NJS_VAR)>([&](__NJS_VAR _needle)
                                                                                              ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/njs.h:877:78: error: expected expression
  function<__NJS_VAR ()>* __OBJ_TO___NJS_STRING = new function<__NJS_VAR ()>([&](){ return __NJS_Create_String((char*)"Array"); });
                                                                             ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp:56:95: error: expected expression
        __NJS_DECL_FUNCTION<__NJS_VAR ()>* __NJS_FN___fc534r = new __NJS_DECL_FUNCTION<__NJS_VAR ()>([&]() -> __NJS_VAR{
                                                                                                     ^
/Users/doberkofler/MyDev/doberkofler/nectarjs-test/.nectar/7urlr/helloworld.cpp:81:94: error: expected expression
__NJS_DECL_FUNCTION<__NJS_VAR ()>* __NJS_FN___2l38vg = new __NJS_DECL_FUNCTION<__NJS_VAR ()>([&]() -> __NJS_VAR{
                                                                                             ^
13 errors generated.

    at checkExecSyncError (child_process.js:611:11)
    at execSync (child_process.js:647:15)
    at Compiler.Compile (/Users/doberkofler/MyDev/doberkofler/nectarjs-test/node_modules/nectarjs/compiler/native/compiler.js:322:3)
    at /Users/doberkofler/MyDev/doberkofler/nectarjs-test/node_modules/nectarjs/nectar.js:351:13
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3) {
  status: 1,
  signal: null,
  output: [
    null,
    <Buffer >,
    <Buffer 49 6e 20 66 69 6c 65 20 69 6e 63 6c 75 64 65 64 20 66 72 6f 6d 20 2f 55 73 65 72 73 2f 64 6f 62 65 72 6b 6f 66 6c 65 72 2f 4d 79 44 65 76 2f 64 6f 62 ... 4200 more bytes>
  ],
  pid: 5012,
  stdout: <Buffer >,
  stderr: <Buffer 49 6e 20 66 69 6c 65 20 69 6e 63 6c 75 64 65 64 20 66 72 6f 6d 20 2f 55 73 65 72 73 2f 64 6f 62 65 72 6b 6f 66 6c 65 72 2f 4d 79 44 65 76 2f 64 6f 62 ... 4200 more bytes>
}
```

**Comment by adrien-thierry** [comment]
Hi @doberkofler ,

Thank you for reporting the issue,

The error comes from Clang, the Apple compiler. I have to investigate because right now NectarJS supports Clang (tested on Windows and Linux).

**Comment by adrien-thierry** [comment]
Commit 5958ea1 should solve the problem,

Could you check it ?

**Comment by XSilent** [comment]
@adrien-thierry I had the exact same problem (on Mac OS X) as @doberkofler. I've just cloned the nectar repository with your recent changes and tried it again and it works!

Btw.:
Super intriguing project!

**Comment by adrien-thierry** [comment]
@XSilent thx for the feedback

The npm version 0.5.23 doesn't solve the pb ?

**Comment by stephenmathieson** [comment]
Just ran a "hello world" myself and saw a similar error.

<details>
<summary>Error details</summary>

```
~/dev/src/github.com/stephenmathieson/test on develop [!?] ∴ node test.js
Hello Stephen

~/dev/src/github.com/stephenmathieson/test on develop [!?] ∴ npx nectar test.js
[*] Generating source file
[*] Compiling
In file included from /Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/test.cpp:42:
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:788:62: error: no matching constructor for initialization of '__NJS_Class_String'
                REGISTER[_array.get().a->__NJS_VALUE[_index]._ptr].s = new __NJS_Class_String(REGISTER[_value._ptr].s->__NJS_VALUE.c_str());
                                                                           ^                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:73:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *') to 'const __NJS_Class_String' for 1st argument
class __NJS_Class_String : public __NJS_Class_Object
      ^
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:73:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *') to '__NJS_Class_String' for 1st argument
class __NJS_Class_String : public __NJS_Class_Object
      ^
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:77:5: note: candidate constructor not viable: 1st argument ('const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *')) would lose const qualifier
    __NJS_Class_String(char* _str);
    ^
1 error generated.
[!] Compilation error
Error: Command failed: g++ -std=c++11 /Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/test.cpp -O1 -fpermissive -w -s   -o /Users/stephen/dev/src/github.com/stephenmathieson/test/test
In file included from /Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/test.cpp:42:
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:788:62: error: no matching constructor for initialization of '__NJS_Class_String'
                REGISTER[_array.get().a->__NJS_VALUE[_index]._ptr].s = new __NJS_Class_String(REGISTER[_value._ptr].s->__NJS_VALUE.c_str());
                                                                           ^                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:73:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *') to 'const __NJS_Class_String' for 1st argument
class __NJS_Class_String : public __NJS_Class_Object
      ^
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:73:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *') to '__NJS_Class_String' for 1st argument
class __NJS_Class_String : public __NJS_Class_Object
      ^
/Users/stephen/dev/src/github.com/stephenmathieson/test/.nectar/xue2m/njs.h:77:5: note: candidate constructor not viable: 1st argument ('const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::value_type *' (aka 'const char *')) would lose const qualifier
    __NJS_Class_String(char* _str);
    ^
1 error generated.

    at checkExecSyncError (child_process.js:630:11)
    at execSync (child_process.js:666:15)
    at Compiler.Compile (/Users/stephen/dev/src/github.com/stephenmathieson/test/node_modules/nectarjs/compiler/native/compiler.js:394:3)
    at /Users/stephen/dev/src/github.com/stephenmathieson/test/node_modules/nectarjs/nectar.js:336:13
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3) {
  status: 1,
  signal: null,
  output: [
    null,
    <Buffer >,
    <Buffer 49 6e 20 66 69 6c 65 20 69 6e 63 6c 75 64 65 64 20 66 72 6f 6d 20 2f 55 73 65 72 73 2f 73 74 65 70 68 65 6e 2f 64 65 76 2f 73 72 63 2f 67 69 74 68 75 ... 1698 more bytes>
  ],
  pid: 27662,
  stdout: <Buffer >,
  stderr: <Buffer 49 6e 20 66 69 6c 65 20 69 6e 63 6c 75 64 65 64 20 66 72 6f 6d 20 2f 55 73 65 72 73 2f 73 74 65 70 68 65 6e 2f 64 65 76 2f 73 72 63 2f 67 69 74 68 75 ... 1698 more bytes>
}

~/dev/src/github.com/stephenmathieson/test on develop [!?] ∴ cat test.js
var name = "Stephen";
console.log("Hello " + name);

~/dev/src/github.com/stephenmathieson/test on develop [!?] ∴
```

</details>

**Comment by adrien-thierry** [comment]
@stephenmathieson it should be good

**Comment by adrien-thierry** [comment]
@stephenmathieson  2c7eccf should solve the problem, can you try it ?

**Comment by adrien-thierry** [comment]
Problem solved for Clang

**Comment by kitolog** [comment]
Hello
I have an error on mac os catalina using nectar

GCC:
`gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin19.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin `

Nectar:
`nectar --version
NectarJS v0.6.61
`

Just tried to compile plain hello.js
`console.log("Hello");`

And got this:

`[*] Generating source file
[*] Compiling
In file included from /Users/macosuser/TestBin/.nectar/yf6e2/hello.cpp:48:
/Users/macosuser/TestBin/.nectar/yf6e2/lib/native_object.h:3:117: error: non-local lambda expression cannot have a capture-default
__NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>* __NJS_FN___qhtdgk = new __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>([=]( vector<var> __NJS_VARARGS ) -> __NJS_VAR{var _obj; if(__NJS_VARARGS.size() > 0) _obj = __NJS_VARARGS[0];
^
    In file included from /Users/macosuser/TestBin/.nectar/yf6e2/hello.cpp:49:
    /Users/macosuser/TestBin/.nectar/yf6e2/lib/json.h:127:117: error: non-local lambda expression cannot have a capture-default
    __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>* __NJS_FN___wn87nk = new __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>([=]( vector<var> __NJS_VARARGS ) -> __NJS_VAR{var __json; if(__NJS_VARARGS.size() > 0) __json = __NJS_VARARGS[0];
    ^
        /Users/macosuser/TestBin/.nectar/yf6e2/lib/json.h:167:119: error: non-local lambda expression cannot have a capture-default
        __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>* __NJS_FN___wczvwh1m = new __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>([=]( vector<var> __NJS_VARARGS ) -> __NJS_VAR{var __object; if(__NJS_VARARGS.size() > 0) __object = __NJS_VARARGS[0];
        ^
            3 errors generated.
                [!] Compilation error
            { Error: Command failed: g++  -std=c++11 /Users/macosuser/TestBin/.nectar/yf6e2/hello.cpp -O1 -fpermissive -w -s   -o /Users/macosuser/TestBin/hello
                In file included from /Users/macosuser/TestBin/.nectar/yf6e2/hello.cpp:48:
                /Users/macosuser/TestBin/.nectar/yf6e2/lib/native_object.h:3:117: error: non-local lambda expression cannot have a capture-default
                __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>* __NJS_FN___qhtdgk = new __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>([=]( vector<var> __NJS_VARARGS ) -> __NJS_VAR{var _obj; if(__NJS_VARARGS.size() > 0) _obj = __NJS_VARARGS[0];
            ^
                In file included from /Users/macosuser/TestBin/.nectar/yf6e2/hello.cpp:49:
                /Users/macosuser/TestBin/.nectar/yf6e2/lib/json.h:127:117: error: non-local lambda expression cannot have a capture-default
                __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>* __NJS_FN___wn87nk = new __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>([=]( vector<var> __NJS_VARARGS ) -> __NJS_VAR{var __json; if(__NJS_VARARGS.size() > 0) __json = __NJS_VARARGS[0];
                ^
                    /Users/macosuser/TestBin/.nectar/yf6e2/lib/json.h:167:119: error: non-local lambda expression cannot have a capture-default
                    __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>* __NJS_FN___wczvwh1m = new __NJS_DECL_FUNCTION<__NJS_VAR (vector<var>)>([=]( vector<var> __NJS_VARARGS ) -> __NJS_VAR{var __object; if(__NJS_VARARGS.size() > 0) __object = __NJS_VARARGS[0];
                    ^
                        3 errors generated.

                            at checkExecSyncError (child_process.js:629:11)
                        at execSync (child_process.js:666:13)
                        at Compiler.Compile (/Users/macosuser/.nvm/versions/node/v10.21.0/lib/node_modules/nectarjs/compiler/native/compiler.js:574:3)
                        at /Users/macosuser/.nvm/versions/node/v10.21.0/lib/node_modules/nectarjs/nectar.js:487:13
                        at FSReqWrap.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:53:3)
                        status: 1,
                            signal: null,
                            output:
                        [ null,
                        <Buffer >,
                    <Buffer 49 6e 20 66 69 6c 65 20 69 6e 63 6c 75 64 65 64 20 66 72 6f 6d 20 2f 55 73 65 72 73 2f 6e 69 6b 69 74 61 2f 50 72 6f 6a 65 63 74 73 2f 53 74 61 72 73 ... > ],
                        pid: 19108,
                            stdout: <Buffer >,
                            stderr:
                    <Buffer 49 6e 20 66 69 6c 65 20 69 6e 63 6c 75 64 65 64 20 66 72 6f 6d 20 2f 55 73 65 72 73 2f 6e 69 6b 69 74 61 2f 50 72 6f 6a 65 63 74 73 2f 53 74 61 72 73 ... > }
`

May you help me?

**Comment by adrien-thierry** [comment]
Hi @kitolog ,

thank you for reporting the issue,

Corrected in Nectar v0.6.62

**Comment by kitolog** [comment]
Right now tested with the new v0.6.62 - all works fine
Thanks!
